File: factory/production_lines/ios/templates/analytics/AnalyticsEvents.swift
// AnalyticsEvents.swift
// DAI-Core Standard Analytics Events
//
// Enum of all standard events every DAI-Core iOS app should track.
// All event names are prefixed with `dai_` automatically.

import Foundation

/// Standard analytics events for DAI-Core iOS apps.
/// Use with `AnalyticsManager.shared.log(_:)`.
enum DAIAnalyticsEvent {

    // MARK: - Session Events

    /// App opened / foregrounded.
    case appOpen
    /// App sent to background.
    case appBackground

    // MARK: - Onboarding Funnel

    /// User started the onboarding flow.
    case onboardingStart
    /// User completed a specific onboarding step.
    case onboardingStep(step: Int, name: String)
    /// User completed the full onboarding.
    case onboardingComplete
    /// User skipped onboarding.
    case onboardingSkip

    // MARK: - Feature Usage

    /// User actively used a feature.
    case featureUsed(name: String)
    /// User discovered / saw a feature for the first time.
    case featureDiscovered(name: String)

    // MARK: - Engagement

    /// Active session duration heartbeat.
    case sessionActive(durationSeconds: Int)
    /// User viewed a piece of content.
    case contentViewed(contentId: String, contentType: String)

    // MARK: - Monetization

    /// User initiated a purchase flow.
    case purchaseStart(productId: String)
    /// Purchase completed successfully.
    case purchaseComplete(productId: String, value: Double, currency: String)
    /// User started a subscription.
    case subscriptionStart(planId: String)
    /// An ad was shown to the user.
    case adImpression(adType: String)

    // MARK: - Errors

    /// A non-fatal error occurred.
    case errorOccurred(domain: String, code: Int, description: String)

    // MARK: - Event Name

    /// Firebase event name, prefixed with `dai_`.
    var name: String {
        let prefix = "dai_"
        switch self {
        // Session
        case .appOpen:
            return "\(prefix)app_open"
        case .appBackground:
            return "\(prefix)app_background"

        // Onboarding
        case .onboardingStart:
            return "\(prefix)onboarding_start"
        case .onboardingStep:
            return "\(prefix)onboarding_step"
        case .onboardingComplete:
            return "\(prefix)onboarding_complete"
        case .onboardingSkip:
            return "\(prefix)onboarding_skip"

        // Feature Usage
        case .featureUsed:
            return "\(prefix)feature_used"
        case .featureDiscovered:
            return "\(prefix)feature_discovered"

        // Engagement
        case .sessionActive:
            return "\(prefix)session_active"
        case .contentViewed:
            return "\(prefix)content_viewed"

        // Monetization
        case .purchaseStart:
            return "\(prefix)purchase_start"
        case .purchaseComplete:
            return "\(prefix)purchase_complete"
        case .subscriptionStart:
            return "\(prefix)subscription_start"
        case .adImpression:
            return "\(prefix)ad_impression"

        // Errors
        case .errorOccurred:
            return "\(prefix)error_occurred"
        }
    }

    // MARK: - Event Parameters

    /// Event-specific parameters dictionary. Returns `nil` for events without parameters.
    var parameters: [String: Any]? {
        switch self {
        // Session — no extra params
        case .appOpen, .appBackground:
            return nil

        // Onboarding
        case .onboardingStart, .onboardingComplete, .onboardingSkip:
            return nil
        case .onboardingStep(let step, let name):
            return [
                "step": step,
                "step_name": name
            ]

        // Feature Usage
        case .featureUsed(let name):
            return ["feature_name": name]
        case .featureDiscovered(let name):
            return ["feature_name": name]

        // Engagement
        case .sessionActive(let durationSeconds):
            return ["duration_seconds": durationSeconds]
        case .contentViewed(let contentId, let contentType):
            return [
                "content_id": contentId,
                "content_type": contentType
            ]

        // Monetization
        case .purchaseStart(let productId):
            return ["product_id": productId]
        case .purchaseComplete(let productId, let value, let currency):
            return [
                "product_id": productId,
                "value": value,
                "currency": currency
            ]
        case .subscriptionStart(let planId):
            return ["plan_id": planId]
        case .adImpression(let adType):
            return ["ad_type": adType]

        // Errors
        case .errorOccurred(let domain, let code, let description):
            return [
                "error_domain": domain,
                "error_code": code,
                "error_description": description
            ]
        }
    }
}

File: factory/production_lines/unity/templates/analytics/AnalyticsEvents.cs
// AnalyticsEvents.cs
// DAI-Core Standard Analytics Events for Unity
//
// Static class with all standard events every DAI-Core Unity app should track.
// Event names are identical to iOS/Android/Web: prefixed with "dai_".
//
// Usage:
//   var (name, parameters) = AnalyticsEvents.AppOpen();
//   AnalyticsManager.Instance.Log(name, parameters);
//
//   var (name, parameters) = AnalyticsEvents.FeatureUsed("dark_mode");
//   AnalyticsManager.Instance.Log(name, parameters);

using System.Collections.Generic;

namespace DriveAI.Analytics
{
    /// <summary>
    /// Standard analytics events for DAI-Core Unity apps.
    /// Static methods return (eventName, parameters) tuples for use with AnalyticsManager.Log().
    /// All event names are pre-prefixed with "dai_" — cross-platform identical.
    /// </summary>
    public static class AnalyticsEvents
    {
        private const string Prefix = "dai_";

        // ── Session Events ───────────────────────────────────────────

        /// <summary>App opened / foregrounded.</summary>
        public static (string name, Dictionary<string, object> parameters) AppOpen()
        {
            return ($"{Prefix}app_open", null);
        }

        /// <summary>App sent to background.</summary>
        public static (string name, Dictionary<string, object> parameters) AppBackground()
        {
            return ($"{Prefix}app_background", null);
        }

        // ── Onboarding Funnel ────────────────────────────────────────

        /// <summary>User started the onboarding flow.</summary>
        public static (string name, Dictionary<string, object> parameters) OnboardingStart()
        {
            return ($"{Prefix}onboarding_start", null);
        }

        /// <summary>User completed a specific onboarding step.</summary>
        public static (string name, Dictionary<string, object> parameters) OnboardingStep(int step, string stepName)
        {
            return ($"{Prefix}onboarding_step", new Dictionary<string, object>
            {
                { "step", step },
                { "step_name", stepName }
            });
        }

        /// <summary>User completed the full onboarding.</summary>
        public static (string name, Dictionary<string, object> parameters) OnboardingComplete()
        {
            return ($"{Prefix}onboarding_complete", null);
        }

        /// <summary>User skipped onboarding.</summary>
        public static (string name, Dictionary<string, object> parameters) OnboardingSkip()
        {
            return ($"{Prefix}onboarding_skip", null);
        }

        // ── Feature Usage ────────────────────────────────────────────

        /// <summary>User actively used a feature.</summary>
        public static (string name, Dictionary<string, object> parameters) FeatureUsed(string featureName)
        {
            return ($"{Prefix}feature_used", new Dictionary<string, object>
            {
                { "feature_name", featureName }
            });
        }

        /// <summary>User discovered / saw a feature for the first time.</summary>
        public static (string name, Dictionary<string, object> parameters) FeatureDiscovered(string featureName)
        {
            return ($"{Prefix}feature_discovered", new Dictionary<string, object>
            {
                { "feature_name", featureName }
            });
        }

        // ── Engagement ───────────────────────────────────────────────

        /// <summary>Active session duration heartbeat.</summary>
        public static (string name, Dictionary<string, object> parameters) SessionActive(int durationSeconds)
        {
            return ($"{Prefix}session_active", new Dictionary<string, object>
            {
                { "duration_seconds", durationSeconds }
            });
        }

        /// <summary>User viewed a piece of content.</summary>
        public static (string name, Dictionary<string, object> parameters) ContentViewed(string contentId, string contentType)
        {
            return ($"{Prefix}content_viewed", new Dictionary<string, object>
            {
                { "content_id", contentId },
                { "content_type", contentType }
            });
        }

        // ── Monetization ─────────────────────────────────────────────

        /// <summary>User initiated a purchase flow.</summary>
        public static (string name, Dictionary<string, object> parameters) PurchaseStart(string productId)
        {
            return ($"{Prefix}purchase_start", new Dictionary<string, object>
            {
                { "product_id", productId }
            });
        }

        /// <summary>Purchase completed successfully.</summary>
        public static (string name, Dictionary<string, object> parameters) PurchaseComplete(string productId, double value, string currency)
        {
            return ($"{Prefix}purchase_complete", new Dictionary<string, object>
            {
                { "product_id", productId },
                { "value", value },
                { "currency", currency }
            });
        }

        /// <summary>User started a subscription.</summary>
        public static (string name, Dictionary<string, object> parameters) SubscriptionStart(string planId)
        {
            return ($"{Prefix}subscription_start", new Dictionary<string, object>
            {
                { "plan_id", planId }
            });
        }

        /// <summary>An ad was shown to the user.</summary>
        public static (string name, Dictionary<string, object> parameters) AdImpression(string adType)
        {
            return ($"{Prefix}ad_impression", new Dictionary<string, object>
            {
                { "ad_type", adType }
            });
        }

        // ── Errors ───────────────────────────────────────────────────

        /// <summary>A non-fatal error occurred.</summary>
        public static (string name, Dictionary<string, object> parameters) ErrorOccurred(string domain, int code, string description)
        {
            return ($"{Prefix}error_occurred", new Dictionary<string, object>
            {
                { "error_domain", domain },
                { "error_code", code },
                { "error_description", description }
            });
        }
    }
}
